lib/semantics/resolve-names.cc

```diff
@@ -4669,6 +4671,51 @@ bool ResolveNamesVisitor::Pre(const parser::ProgramUnit &x) {
   return false;
 }
 
+// Calls to dummy procedures need to record that their symbols are known
+// to be procedures, so that they don't get converted to objects by default.
+class ExecutionPartSkimmer {
+public:
+  ExecutionPartSkimmer(SemanticsContext &c, Scope &s)
+    : context_{c}, scope_{s} {}
+
+  void Walk(const parser::ExecutionPart *exec) {
+    if (exec != nullptr) {
+      parser::Walk(*exec, *this);
+    }
+  }
+
+  template<typename A> bool Pre(const A &) { return true; }
+  template<typename A> void Post(const A &) {}
+  void Post(const parser::FunctionReference &fr) {
+    NoteCall(Symbol::Flag::Function, fr.v);
+  }
+  void Post(const parser::CallStmt &cs) {
+    NoteCall(Symbol::Flag::Subroutine, cs.v);
+  }
+
+private:
+  void NoteCall(Symbol::Flag, const parser::Call &);
+
+  SemanticsContext &context_;
+  Scope &scope_;
+};
+
+void ExecutionPartSkimmer::NoteCall(
+    Symbol::Flag flag, const parser::Call &call) {
+  auto &designator{std::get<parser::ProcedureDesignator>(call.t)};
+  if (const auto *name{std::get_if<parser::Name>(&designator.u)}) {
+    if (Symbol * symbol{scope_.FindSymbol(name->source)}) {
```

Comment:
`FindSymbol` could find the symbol in an outer subprogram if it's not in `scope_`. It could be that doesn't matter because it will never have `EntityDetails`. But it might be safer to use `scope_.find(name->source)` instead so that it looks only in `scope_`.

Reply:
A host's dummy procedure can be called from a contained subprogram, and that might be the only call, so the skimming of the contained subprogram's executable part needs to be able to access its host's scope.

Reply:
OK